src/lib/utils.py

```diff
@@ -425,13 +434,19 @@ def dump_manifest(contents, manifest_path):
     """Writes back 'contents' to 'manifest_path'.
 
     For YAML, we make a best-effort attempt to preserve
-    formatting; for JSON, we use the canonical 4-space indentation."""
+    formatting; for JSON, we use the canonical 4-space indentation,
+    but add a trailing newline if originally present."""
     _, ext = os.path.splitext(manifest_path)
-    with open(manifest_path, "w", encoding="utf-8") as fp:
+    with open(manifest_path, "r+", encoding="utf-8") as fp:
         if ext in (".yaml", ".yml"):
             _yaml.dump(contents, fp)
         else:
+            newline = _check_newline(fp)
+            fp.seek(0, 0)
```

Comment:
Nitpick: remember and return to the original position in `_check_newline` instead resetting it outside.

Reply:
I agree, that would be less ugly.